middleware/throttle.go

```diff
@@ -34,21 +54,46 @@ type token struct{}
 
 // throttler limits number of currently processed requests at a time.
 type throttle struct {
-	h      chi.Handler
-	tokens chan token
+	h             chi.Handler
+	tokens        chan token
+	backlogtokens chan token
+	timeout       time.Duration
 }
 
 // TODO: add support for a backlog
 
 // ServeHTTPC implements chi.Handler interface.
 func (t *throttle) ServeHTTPC(ctx context.Context, w http.ResponseWriter, r *http.Request) {
+	timer := time.NewTimer(t.timeout)
+
 	select {
+	case <-timer.C:
```

Comment:
Thinking of the behavior...

Imho, if the backlog is full, we should return Service Unavailable response as soon as possible. Only if the request can fit into the backlog, then we should start the timeout -- while waiting for the throttle token.

So, I'd remove timer from this select and move the initialization down:

```diff
- case btok := <-t.backlogtokens:
+ case btok := <-t.backlogtokens:
+    timer := time.NewTimer(t.timeout)
      // ... case for time.C, ctx.Done() and t.tokens
+ default:
+    httpStatus(w, http.StatusServiceUnavailable)
+    return
```